components/builder-worker/src/runner/mod.rs

```diff
@@ -0,0 +1,350 @@
+/// Protocol message to indicate the Job Runner has completed a work request
+const WORK_COMPLETE: &'static str = "C";
+
+lazy_static! {
+    // JW TODO: expose public API functions in the core crate to check if the Rust proccess which
+    // is currently executing is, itself, packaged by Habitat. If so, then we should expose another
+    // public API function to load the PackageInstall of a dep for the given `origin`/`name`
+    // combination. If we can't, then we should fall back to the latest of core/hab-studio because
+    // that means we're just in a dev shell.
+    static ref STUDIO_PKG: PackageIdent = PackageIdent::from_str("core/hab-studio").unwrap();
+}
+
+pub struct Runner {
+    pub workspace: Workspace,
+    auth_token: String,
+}
+
+impl Runner {
+    pub fn new(job: proto::Job, config: &Config) -> Self {
+        Runner {
+            auth_token: config.auth_token.clone(),
+            workspace: Workspace::new(config.data_path.clone(), job),
+        }
+    }
+
+    pub fn run(&mut self) -> () {
+        if let Some(err) = self.setup().err() {
+            error!("WORKSPACE SETUP ERR={:?}", err);
+            return self.fail();
+        }
+        let mut logger = Logger::new(&self.workspace);
+        // JW TODO: How are we going to get the secret keys for this thing?
+        if let Some(err) = vcs::clone(&self.workspace.job, &self.workspace.src()).err() {
+            error!("CLONE ERROR={}", err);
+            return self.fail();
+        }
+        let mut archive = match self.build(&mut logger) {
+            Ok(archive) => archive,
+            Err(err) => {
+                error!("STUDIO ERR={}", err);
+                return self.fail();
+            }
+        };
+        if !self.post_process(&mut archive) {
+            // JW TODO: We should shelve the built artifacts and allow a retry on post-processing.
+            // If the job is killed then we can kill the shelved artifacts.
+            return self.fail();
+        }
+        self.complete()
+    }
+
+    pub fn build(&self, logger: &mut Logger) -> Result<PackageArchive> {
+        let args =
+            vec![OsString::from("-s"),
+                 OsString::from(self.workspace.src()),
+                 OsString::from("-r"),
+                 OsString::from(self.workspace.studio()),
+                 OsString::from("-k"),
+                 OsString::from("core"),
+                 OsString::from("build"),
+                 OsString::from(Path::new(self.workspace.job.get_project().get_plan_path())
+                     .parent()
+                     .unwrap())];
+        let command = match PackageInstall::load(&STUDIO_PKG, None) {
+            Ok(package) => format!("{}/hab-studio", package.paths().unwrap()[0].display()),
+            Err(_) => panic!("should install core/hab-studio because it's missing?"),
+        };
+        debug!("building, cmd={:?}, args={:?}", command, args);
+        let mut child = Command::new(command)
```

Comment:
It makes me so happy to see our uses of unsafe libc calls pushed back to standard library functions. 👍 Go std::process::Command!